src/sources/metadatasourcetaglib.cpp

```diff
@@ -629,22 +629,53 @@ class AiffTagSaver: public TagSaver {
  */
 class SafelyWritableFile final {
   public:
-    SafelyWritableFile(QString origFileName, bool useTemporaryFile)
-        : m_origFileName(std::move(origFileName)) {
+    SafelyWritableFile(QString origFileName, bool useTemporaryFile) {
+        // Both file names remain uninitialized until all prerequisite operations
+        // in the constructor have been completed successfully. Otherwise failure
+        // to create the temporary file will not be handled correctly!
+        // See also: https://bugs.launchpad.net/mixxx/+bug/1815305
+        DEBUG_ASSERT(m_origFileName.isNull());
         DEBUG_ASSERT(m_tempFileName.isNull());
         if (useTemporaryFile) {
-            QString tempFileName = m_origFileName + kSafelyWritableTempFileSuffix;
-            QFile origFile(m_origFileName);
-            if (origFile.copy(tempFileName)) {
-                m_tempFileName = std::move(tempFileName);
-            } else {
+            QString tempFileName = origFileName + kSafelyWritableTempFileSuffix;
+            QFile origFile(origFileName);
+            if (!origFile.copy(tempFileName)) {
                 kLogger.warning()
                         << origFile.errorString()
-                        << "- Failed to copy original into temporary file before writing:"
-                        << origFile.fileName()
+                        << "- Failed to clone original into temporary file before writing:"
+                        << origFileName
                         << "->"
                         << tempFileName;
+                // Abort constructor
+                return;
+            }
+            QFile tempFile(tempFileName);
+            DEBUG_ASSERT(tempFile.exists());
+            // Both file sizes are expected to be equal after successfully
+            // copying the file contents.
+            VERIFY_OR_DEBUG_ASSERT(origFile.size() == tempFile.size()) {
+                kLogger.warning()
+                        << "Failed to verify size after cloning original into temporary file before writing:"
+                        << origFile.size()
+                        << "<>"
+                        << tempFile.size();
+                // Cleanup
+                if (tempFile.exists() && !tempFile.remove()) {
+                    kLogger.warning()
+                            << tempFile.errorString()
+                            << "- Failed to remove temporary file:"
+                            << tempFileName;
+                }
+                // Abort constructor
+                return;
             }
+            // Successfully cloned original into temporary file for writing - finish initialization
+            m_origFileName = std::move(origFileName);
+            m_tempFileName = std::move(tempFileName);
+        } else {
+            // Directly write into original file - finish initialization
```

Comment:
Can/should we do a free-space check here in case the new meta-data is larger?

Reply:
No, we cannot make any sensible assumptions here. That's the point why we write into a temporary file instead of modifying the original file. It doesn't matter if that fails, because we can safely perform a rollback.